### PR TITLE
Add CloseableTracer utility class

### PR DESCRIPTION
**Pull request by raiju**
Adds a convenience class to make adding tracers cleaner.

@j-baker @uschi2000

### CODE REVIEW

**Comment by raiju** [comment]
The test failure looks unrelated to this PR

File: tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * Wraps the {@link Tracer} methods in a closeable resource to enable the usage of the try-with-resources pattern.
+ *
+ * Usage:
+ *   try (CloseableTracer trace = CloseableTracer.start("traceName")) {
+ *       [...]
+ *   }
+ */
+public final class CloseableTracer implements AutoCloseable {
```
Comment (j-baker) [inline tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java]: maybe just have this be an enum?

Comment (raiju) [reply tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java]: What is the advantage?

Comment (raiju) [reply tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java]: Thinking about it some more, I'm not 100% sure on what you mean. Do you mean something like:
```
enum CloseableTracer implements AutoCloseable {
  INSTANCE;
  static CloseableTracer startSpan(...) { ... }
}
```
?

In that case what would stop the user from accessing `CloseableTracer.INSTANCE` directly? It's also not obvious to me what the advantage would be. Did I misunderstand what you meant?

Comment (j-baker) [reply tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java]: oh, sorry. So I was just suggesting that given that you only want there to be one of these, one thing which people frequently end up doing is making it an enum. But I don't really care and don't want to block this.

**Review by j-baker — APPROVED** [review]

File: tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+import com.palantir.remoting.api.tracing.SpanType;
+
+/**
+ * Wraps the {@link Tracer} methods in a closeable resource to enable the usage of the try-with-resources pattern.
+ *
+ * Usage:
+ *   try (CloseableTracer trace = CloseableTracer.start("traceName")) {
+ *       [...]
+ *   }
```
Comment (iamdanfox) [inline tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java:27]: Hey @raiju any chance we could also add one sentence of guidance so people know roughly when they should use this? The main thing I want to convey is that http-remoting calls are already traced, so people don't need to wrap every single call they ever do in this!

**Review by iamdanfox — APPROVED** [review]
Approve pending a tiny bit of Javadoc

File: tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    public static CloseableTracer startSpan(String operation) {
+        Tracer.startSpan(operation);
+        return INSTANCE;
+    }
+
+    /**
+     * Like {@link #startSpan(String)}, but opens a span of the explicitly given {@link SpanType span type}.
+     */
+    public static CloseableTracer startSpan(String operation, SpanType spanType) {
```
Comment (iamdanfox) [inline tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java]: `SpanType` allows SERVER_INCOMING, CLIENT_OUTGOING or LOCAL... but we're only ever expecting this to be called with CLIENT_OUTGOING or LOCAL right?

Comment (j-baker) [reply tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java]: probably only LOCAL, here. CLIENT_OUTGOING one would anticipate should be implemented inside the client.

Comment (iamdanfox) [reply tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java]: Hmm, so maybe the single-arg variant would be sufficient then?  `public static CloseableTracer startSpan(String operation)` just does a LOCAL span

Comment (raiju) [reply tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java]: I'm not sure why we'd limit it; the current APIs (`Tracer.[...]`) allow all these calls.

File: tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java (hunk 2 above)
Comment (raiju) [reply tracing/src/main/java/com/palantir/remoting3/tracing/CloseableTracer.java:27]: Is this class that different from https://github.com/palantir/http-remoting/blob/29b83d50358d17e8f57eca62b2790ec6e08ecbf6/tracing/src/main/java/com/palantir/remoting3/tracing/Tracer.java#L42 to warrant this?

**Comment by uschi2000** [comment]
Please add test coverage for these kind of changes, even if it's trivial. https://github.com/palantir/http-remoting/pull/718

**Comment by raiju** [comment]
@uschi2000 My bad, thanks for adding!

**Pull request merged**